File: intern/143.reorder-list.js
/*
 * @lc app=leetcode id=143 lang=javascript
 *
 * [143] Reorder List
 */

/**
 * @param {number} val
 * @param {ListNode|null} next
 */
function ListNode(val, next) {
    this.val = val;
    this.next = next;
}

// @lc code=start
/**
 * @param {ListNode} head
 * @return {void} Do not return anything, modify head in-place instead.
 */
const reorderList = function (head) {
    if (head.next === null) {
        return;
    }

    /** @type {ListNode|null} */
    let fast = head;
    /** @type {ListNode|null} */
    let slow = head;
    let beforeSlow = head;
    while (true) {
        beforeSlow = slow;
        slow = slow.next;
        fast = fast.next;
        if (fast === null) {
            break;
        }
        fast = fast.next;
        if (fast === null) {
            break;
        }
    }
    // slow 是逆序的起始点
    beforeSlow.next = null;

    /** @type {ListNode[]} */
    const secondHalfStack = [];
    while (slow !== null) {
        secondHalfStack.push(slow);
        slow = slow.next;
    }

    let currentNode = head;
    let currentNodeNext = head.next;
    while (secondHalfStack.length > 0) {
        const listNodeToInsert = secondHalfStack.pop();
        currentNode.next = listNodeToInsert;
        listNodeToInsert.next = currentNodeNext;
        currentNode = currentNodeNext;
        if (currentNode !== null) {
            currentNodeNext = currentNode.next;
        }
    }
};
// @lc code=end

File: src/2023-2024/143.reorder-list.cpp
/*
 * @lc app=leetcode id=143 lang=cpp
 *
 * [143] Reorder List
 */
struct ListNode {
  int val;
  ListNode* next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(next) {}
};
// @lc code=start
class Solution {
 public:
  void reorderList(ListNode* head) {
    ListNode* firstPartTail = getFirstPartTail(head);
    ListNode* secondPartHead = firstPartTail->next;
    firstPartTail->next = nullptr;
    secondPartHead = reverseList(secondPartHead);

    alternateMergeLists(head, secondPartHead);
  }

  ListNode* reverseList(ListNode* head) {
    ListNode fakeHead;
    ListNode* currentNode = head;
    while (currentNode != nullptr) {
      ListNode* nextNode = currentNode->next;
      currentNode->next = fakeHead.next;
      fakeHead.next = currentNode;
      currentNode = nextNode;
    }
    return fakeHead.next;
  }

  ListNode* getFirstPartTail(ListNode* head) {
    ListNode* prevSlow = nullptr;
    ListNode* slow = head;
    ListNode* fast = head;
    while (fast != nullptr && fast->next != nullptr) {
      prevSlow = slow;
      slow = slow->next;
      fast = fast->next->next;
    }

    if (fast != nullptr) {
      return slow;
    }

    return prevSlow;
  }

  ListNode* alternateMergeLists(ListNode* head1, ListNode* head2) {
    ListNode* currentList1Node = head1;
    ListNode* currentList2Node = head2;

    while (currentList2Node != nullptr) {
      auto nextList1Node = currentList1Node->next;
      auto nextList2Node = currentList2Node->next;

      currentList1Node->next = currentList2Node;
      currentList2Node->next = nextList1Node;

      currentList1Node = nextList1Node;
      currentList2Node = nextList2Node;
    }

    return head1;
  }
};
// @lc code=end
